### DISCUSSION TITLE
Ball devices without ball_switches

### DISCUSSION THREAD

**Discussion by worldpeace-germany** [missionpinball/missionpinball#52] [Q&A]
Hello,

for a homebrew solution I would like to use MPF to run without ball_switches in my ball devices. The idea is the following:

- When a player presses the start button, score is set to zero and they get 3 balls
- A ball is taken out of a trough, which has no switches to recognize how many balls are in there
- The ball is placed by hand in front of the plunger, again no switch so MPF has still no clue where the ball is
- The ball is fired onto the playfield from the mechnical plunger, MPF recognizes the ball is active via a switch tag `tags: playfield_active`
- The trough has an entrance_switch configured to recognize that a ball is lost

I tried to set this up virtually (forget about the score, it is only about ball devices for now) based on the tutorial step 10. See relevant config file at the bottom.

The first problem I hit is when I press the start button I am not exiting the attract mode. It seems that this only works when my ball devices shows via a switch that there is a ball somewhere.

Any idea or hint what I should do to achieve this?

Thanks!

```
#config_version=5

hardware:
  platform: smart_virtual

switches:
  s_trough_entrance:
    number: 1
  s_start:
    number: 11
    tags: start
  s_right_inlane:
    number: 12
    tags: playfield_active

keyboard:
  s:
    switch: s_start
  q:
    switch: s_right_inlane
  1:
    switch: s_trough_entrance


ball_devices:
  bd_trough:
    entrance_switch: s_trough_entrance # no effect if ball_switches is set
    ball_capacity: 3
    mechanical_eject: true
    tags: trough, home, drain
    eject_targets: bd_plunger
    debug: true
  bd_plunger:
    ball_capacity: 1
    mechanical_eject: true

playfields:
  playfield:
    tags: default
    default_source_device: bd_plunger
```

**Comment by worldpeace-germany** [comment]
I got one step further, I found this setting


```
machine:
  balls_installed: 3 
  min_balls: 0
```
with min_balls=0 I am able to start the game and go from attract mode into game mode. But when I press "q" to show that my playfield is active, no ball is shown on the playfield.

Am I moving into the right direction? If yes any idea how to take the next step?

**Reply by Stefku** [reply]
Just because I'm curious. Why don't you put a switch at the plunger? Are there restrictions of the hardware?

**Reply by worldpeace-germany** [reply]
No restrictions, it is for a homebrew project. The simple reason is things you don't use cannot break, they don't need maintenance and no adjustments needed. I hope I don't need such a switch to trick MPF. To trick MPF I would rather try to have some virtual switches as ball_switches in my trough and show them always as active.

**Reply by Stefku** [reply]
ok, I understand. How about trying to remote control the plunger lane switch. The following worked for me
With the following the left flipper is remote controlled after 500ms after the right flipper has been switched.
```
event_player:
    s_right_flipper_active: s_left_flipper_active|500ms
```
I also tried to remote control the plunger lane switch but it did not work. But I didn't invest much time ;)
```
event_player:
  s_right_flipper_active: s_plunger_lane_active|500ms
  s_plunger_lane_active: s_plunger_lane_inactive|500ms #do not lead to "ball device - ball left"
```

**Reply by worldpeace-germany** [reply]
Thanks for that idea. Doesn't seem to work for me. I see that the event (in your naming) s_left_flipper_active is being fired. But nothing else is happening. Could it be that there is more if a switch is closed than sending that event? Or in other words sending that event doesn't simulate fully the use of a switch? Is your left flipper physically moving after you pressed the right flipper?

**Reply by Stefku** [reply]
Actually, I tested it in virtual mode "-X". And there at least some other events were fired that were bound to the flipper.

I played around a little more and found out that you can get rid off the plunger with defining that the trough releases the ball directly to the playfield. That reduces a bit the problem. But, I could not remote control the switch for the ball in the trough. If you are a bit familiar with python you could try to remote control the trough with custom code: https://developer.missionpinball.org/en/dev/code/machine_code.html and https://developer.missionpinball.org/en/dev/api/self.machine.ball_devices.html.

Or, just add one switch for the balls in the trough :)

**Comment by worldpeace-germany** [comment]
Python itself should not be a problem for me. Though I wondering if that is really needed and I would like to avoid that, again just maintenance.  I need to think about the logic I need to implement. But for now that seems my last option. Apart from that

- You said you used the -X switch, I have in my config file platform: smart_virtual from what I understand it should be the same as -X? Or am I misunderstanding something?

- I would not mind skipping the plunger as a ball_device.  I tried that as well before, but again no luck. Not getting it to work.

- I tried as well to have in the trough one switch which is always activated. Again no luck.

**Reply by Stefku** [reply]
Yes, -X is the smart_virtual platform. But it's just virtual, thus you cannot switch real hardware in that mode.
BTW: What hardware do you want to use? Electronic just for the switches? Also high voltage for coils?

As far as I understand, MPF is designed in a way, that it always knows where the balls are ([see Ball Tracking](https://docs.missionpinball.org/en/stable/game_logic/ball_tracking/index.html)) The only "sense" MPF has are switches. Thus I guess your use case is not the common and maybe not implemented. But I don't know all the details of the platform.

Theoretical thoughts. You need a (new kind of?) ball device with just one switch for the entry. With playfields switches the ball device infers that the ball left the ball device. Another consequence is, that you have a '1 ball game' only, because the playfield alone is not able to count the balls.

**Reply by worldpeace-germany** [reply]
Hi,

yes ball tracking is important, but from what I understood the ball_capacity is exactly for the use case where you don't have ball_switches. I agree, my use case is maybe not the most common one and thus it might not be fully tested/implemented. At the end of the day my hardware I going to use is the Cobra OPP board.

My complete current test config looks like this

```
#config_version=5

hardware:
  platform: smart_virtua
  
switches:
  s_trough_entrance:
    number: 1
  s_start:
    debug: true
    number: 11
    tags: start
  s_right_inlane:
    number: 12
    tags: playfield_active
  s_plunger:
    debug: true
    number: 10

keyboard:
  s:
    switch: s_start
  q:
    switch: s_right_inlane
  1:
    switch: s_trough_entrance
  p:
    switch: s_plunger
    toggle: true


ball_devices:
  bd_trough:
    entrance_switch: s_trough_entrance
    ball_capacity: 3
    mechanical_eject: true
    tags: trough, home, drain
    eject_targets: bd_plunger
    debug: true
  bd_plunger:
    #ball_capacity: 1
    ball_switches: s_plunger
    mechanical_eject: true

playfields:
  playfield:
    tags: default
    default_source_device: bd_plunger
    
machine:
  balls_installed: 3 #How many balls are physically in your game
  min_balls: 0 #How few balls can be accounted for before you can start a game


game:
  balls_per_game: 3
  max_players: 4

displays:
  window:
    width: 800
    height: 600

slides:
  welcome_slide:
    widgets:
      - type: text
        text: PINBALL!
        font_size: 50
        color: red
      - type: rectangle
        width: 240
        height: 60
  attract_started:
    widgets:
      - text: ATTRACT MODE
        type: text

slide_player:
  init_done: welcome_slide
  mode_attract_started: attract_started
  ball_started:
    widgets:
      type: text
      text: PLAYER (number) BALL (ball)
```

I can do

- press s to start the game
- press p to have the ball in the plunger
- press p to have the ball leaving the plunger
- press q to show that the ball is active on the playfield
- press 1 to show that the ball drained

That is basically working with the following open questions/issue

a) I want to get rid of the "p" switch, I am not able to remove the p switch and have it working the same way.
b) balls are not counted correctly, even the second ball shows that it is ball no. 1 - a bug in the entrance_switch usage? (balls_per_game is set to 3, so the entrance_switch should be able to count down - I would assume/hope)

**Comment by worldpeace-germany** [comment]
closing it, very old and not valid anymore. By now i use a switch in the plunger.
